File: fxgl-entity/src/main/java/com/almasb/fxgl/entity/components/CollidableComponent.java
/*
 * FXGL - JavaFX Game Library. The MIT License (MIT).
 * Copyright (c) AlmasB (almaslvl@gmail.com).
 * See LICENSE for details.
 */

package com.almasb.fxgl.entity.components;

import com.almasb.fxgl.entity.component.CopyableComponent;
import com.almasb.fxgl.core.util.BackportKt;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

/**
 * Marks an entity as collidable.
 *
 * @author Almas Baimagambetov (AlmasB) (almaslvl@gmail.com)
 */
public class CollidableComponent extends BooleanComponent implements CopyableComponent<CollidableComponent> {

    private List<Serializable> ignoredTypes = new ArrayList<>();

    public CollidableComponent(boolean collidable) {
        super(collidable);
    }

    public List<Serializable> getIgnoredTypes() {
        return ignoredTypes;
    }

    public void addIgnoredType(Serializable type) {
        ignoredTypes.add(type);
    }

    public void removeIgnoredType(Serializable type) {
        ignoredTypes.remove(type);
    }

    @Override
    public CollidableComponent copy() {
        CollidableComponent cc = new CollidableComponent(getValue());
        BackportKt.forEach(ignoredTypes, cc::addIgnoredType);

        return cc;
    }
}

File: fxgl-entity/src/main/kotlin/com/almasb/fxgl/entity/components/CollidableComponent.kt
/*
 * FXGL - JavaFX Game Library. The MIT License (MIT).
 * Copyright (c) AlmasB (almaslvl@gmail.com).
 * See LICENSE for details.
 */

package com.almasb.fxgl.entity.components

import com.almasb.fxgl.entity.component.CopyableComponent
import java.io.Serializable
import java.util.*

/**
 * Marks an entity as collidable.
 *
 * @author Almas Baimagambetov (AlmasB) (almaslvl@gmail.com)
 */
class CollidableComponent(collidable: Boolean) : BooleanComponent(collidable), CopyableComponent<CollidableComponent> {

    private val ignoredTypes = ArrayList<Serializable>()

    fun getIgnoredTypes(): List<Serializable> {
        return ignoredTypes
    }

    fun addIgnoredType(type: Serializable) {
        ignoredTypes.add(type)
    }

    fun removeIgnoredType(type: Serializable) {
        ignoredTypes.remove(type)
    }

    override fun copy(): CollidableComponent {
        return CollidableComponent(value).also { it.ignoredTypes.addAll(ignoredTypes) }
    }

    override fun isComponentInjectionRequired(): Boolean = false
}